### PR TITLE
tighten the checks for `Trainer.terminate_on_nan`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #9189

- Moves the check for infinite loss earlier in the training loop right after the loss is returned from the `LightningModule.training_step`. 
- Fixes also a utility function which called also the nan detection for model parameters when it was only meant to check the loss value.
- Adds an assert for the expected number of backward calls when terminating

**Context:** `Trainer.terminate_on_nan = True` makes two checks:
- check loss value for NaN or inf before we do the backward and terminate if it is the case. 
- check parameter values after backward, terminate if it is the case.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -266,9 +263,9 @@ def _make_backward_fn(self, optimizer: Optimizer, opt_idx: int) -> Optional[Call
         def backward_fn(loss: Tensor):
             self.backward(loss, optimizer, opt_idx)
 
-            # check if loss or model weights are nan
             if self.trainer.terminate_on_nan:
-                check_finite_loss(self.trainer.lightning_module, loss)
+                # check if model weights are nan after backward
+                detect_nan_parameters(self.trainer.lightning_module)
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:268]: reviewer: checking the loss value only makes sense before backward. 
here we are only interested in checking the parameter values of the model.

File: pytorch_lightning/loops/utilities.py [hunk 2]
```diff
@@ -100,6 +97,9 @@ def _process_training_step_output(
     # map to results under the hood
     results.minimize = loss
 
+    if trainer.terminate_on_nan:
+        check_finite_loss(loss)
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:102]: reviewer: the check for the loss gets moved here, which happens when we get loss and results from the `LightningModule.training_step`

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -808,6 +810,7 @@ def training_step(self, batch, batch_idx):
     with pytest.raises(ValueError, match=r".*The loss returned in `training_step` is.*"):
         trainer.fit(model)
         assert trainer.global_step == model.test_batch_inf
+        assert backward_mock.call_count == model.test_batch_inf
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:813]: reviewer: the test here ensures we are terminating before the 4th backward pass.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/utilities.py [hunk 4]
```diff
@@ -25,21 +25,18 @@
...
-    """Checks for finite parameters and loss values.
+def check_finite_loss(loss: torch.Tensor) -> None:
+    """Checks for finite loss value.
 
     Args:
-        model: a reference to the ``LightningModule``
         loss: the loss value to check to be finite
     """
-    if not torch.isfinite(loss).all():
+    if loss is not None and not torch.isfinite(loss).all():
```
Comment (ananthsub) [inline pytorch_lightning/loops/utilities.py:38]: is this needed given the loss is expected to be a Tensor? when do we pass None as an input?

File: pytorch_lightning/loops/utilities.py [hunk 5]
```diff
@@ -25,21 +25,18 @@
...
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from pytorch_lightning.utilities.fetching import AbstractDataFetcher, DataLoaderIterDataFetcher
-from pytorch_lightning.utilities.finite_checks import detect_nan_parameters
 from pytorch_lightning.utilities.signature_utils import is_param_in_hook_signature
 from pytorch_lightning.utilities.types import STEP_OUTPUT
 
 
-def check_finite_loss(model: "pl.LightningModule", loss: torch.Tensor) -> None:
-    """Checks for finite parameters and loss values.
+def check_finite_loss(loss: torch.Tensor) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/loops/utilities.py:32]: technically this is a breaking change. could we make a separate folder for internal utilities intended for only the trainer or loops to use?

Comment (awaelchli) [reply pytorch_lightning/loops/utilities.py:32]: how so? this was never introduced in any release. 
and this already is in a separate folder just for loops.

Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:32]:
```suggestion
def check_finite_loss(loss: Optional[torch.Tensor]) -> None:
```

File: pytorch_lightning/loops/utilities.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/loops/utilities.py:38]: I can update the type hint. 
None comes in when the training_step does not return a loss (e.g. in manual opt or when skipping a step). 

This will anyway be removed once deprecation phase of #8313 is over

**Review by ananthsub — APPROVED** [review]
